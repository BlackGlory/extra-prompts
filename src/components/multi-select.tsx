import { Box, Text } from 'ink'
import { MultiSelect as MultiSelectInput } from '@inkjs/ui'
import { IOption } from '@src/types.js'

interface IMultipleSelectOptions<T> {
  message: string
  options: IOption<T>[]

  onSelect: (options: Array<IOption<T>>) => void
}

export function MultiSelect<T>(props: IMultipleSelectOptions<T>) {
  const { message, options, onSelect } = props

  return (
    <Box>
      <Text>{message}: </Text>
      <MultiSelectInput
        options={options.map((option, i) => ({
          label: option.label
        , value: i.toString()
        }))}
        onSubmit={indexStrings => {
          const indexes = indexStrings.map(indexString => Number(indexString))
          const option = indexes.map(index => options[index])

          onSelect(option)
        }}
      />
    </Box>
  )
}
