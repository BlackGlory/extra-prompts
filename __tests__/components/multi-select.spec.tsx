import { MultiSelect } from '@components/multi-select.js'
import { render } from 'ink-testing-library'
import { delay } from 'extra-promise'
import { expect } from 'vitest'
import { stripVTControlCharacters } from 'util'

expect.addSnapshotSerializer({
  serialize: stripVTControlCharacters
, test: () => true
})

describe('MultiSelect', () => {
  test('render', () => {
    const fn = vi.fn()
    const { lastFrame } = render(
      <MultiSelect
        message='message'
        onSelect={fn}
        options={[
          { label: 'Option 1', value: 1 }
        , { label: 'Option 2', value: 2 }
        , { label: 'Option 3', value: 3 }
        ]}
      />
    )

    expect(lastFrame()).toMatchSnapshot()
    expect(fn).toBeCalledTimes(0)
  })

  test('onSelect', async () => {
    const fn = vi.fn()
    const { lastFrame, stdin } = render(
      <MultiSelect
        message='message'
        onSelect={fn}
        options={[
          { label: 'Option 1', value: 1 }
        , { label: 'Option 2', value: 2 }
        , { label: 'Option 3', value: 3 }
        ]}
      />
    )

    await delay(100)
    stdin.write('\u001b[B') // down
    await delay(100)
    stdin.write(' ')
    await delay(100)
    stdin.write('\u001b[B') // down
    await delay(100)
    stdin.write(' ')
    await delay(100)
    stdin.write('\r')
    await delay(100)

    expect(lastFrame()).toMatchSnapshot()
    expect(fn).toBeCalledTimes(1)
    expect(fn).toBeCalledWith([
      {
        label: 'Option 2'
      , value: 2
      }
    , {
        label: 'Option 3'
      , value: 3
      }
    ])
  })
})
