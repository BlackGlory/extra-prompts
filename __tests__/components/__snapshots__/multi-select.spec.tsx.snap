// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiSelect > onSelect 1`] = `
message:   Option 1
           Option 2 ✔
         ❯ Option 3 ✔
`;

exports[`MultiSelect > render 1`] = `
message: ❯ Option 1
           Option 2
           Option 3
`;
